turbo/snapshotsync/block_snapshots.go

```diff
@@ -491,17 +499,25 @@ func (s *RoSnapshots) Reopen() error {
 
 	return nil
 }
-func (s *RoSnapshots) ReopenSegments() error {
+func (s *RoSnapshots) ReopenSegments(ctx context.Context, snapshotDownloader proto_downloader.DownloaderClient) error {
 	s.Headers.lock.Lock()
 	defer s.Headers.lock.Unlock()
 	s.Bodies.lock.Lock()
 	defer s.Bodies.lock.Unlock()
 	s.Txs.lock.Lock()
 	defer s.Txs.lock.Unlock()
 	s.closeSegmentsLocked()
-	files, err := segments(s.dir)
+	files, missingSnapshots, err := segments(s.dir)
 	if err != nil {
-		return err
+		// if we have missing snapshots then the error is a missing snapshots error
+		// so we attempt to download the snapshots
+		if len(missingSnapshots) > 0 && snapshotDownloader != nil {
+			if err := requestSnapshotDownload(ctx, missingSnapshots, snapshotDownloader); err != nil {
```

Comment:
Mess network-related code with other code usually bad practice, because:
- “ network-related code” it’s “control flow” - it must be easy to understand the flow and draw architecture diagram: which service calling which service. After this change “any service which open snapshots for read - may pass downloader client and initiate downloading of 500Gb”
- “network-related code” is slow. It does many retries if network is broken or on timeout, downloader service may not start yet - it may start after 10minutes. Means now it’s unclear “method Reopen() is fast or slow?” - usually it’s fast. And here you think that “it’s fast”, do lock global mutex and do network communication while mutex (and whole other app) is locked. Other people will do same mistake, because they will feel “this method is usually fast”. 
- do bind logic of app to “nil-ness of variable” it’s also bad practice - because “anything in any time”  can become nil (it’s unfriendly and unsafe environment). After adding more layers of abstractions here: other dev will not able understand “what will happen if i pass nil here” without reading all source code of underlying abstractions. And also they may think “i’m not sure why, but this method asking Downloader as a parameter, i will pass it here jut to be safe”.